src/buffer/out/textBuffer.cpp

```diff
@@ -2888,125 +2900,400 @@ std::vector<til::point_span> TextBuffer::SearchText(const std::wstring_view& nee
     return results;
 }
 
-const std::vector<ScrollMark>& TextBuffer::GetMarks() const noexcept
+// Collect up all the rows that were marked, and the data marked on that row.
+// This is what should be used for hot paths, like updating the scrollbar.
+std::vector<ScrollMark> TextBuffer::GetMarkRows() const
+{
+    std::vector<ScrollMark> marks;
+    const auto bottom = _estimateOffsetOfLastCommittedRow();
+    for (auto y = 0; y <= bottom; y++)
+    {
+        const auto& row = GetRowByOffset(y);
+        const auto& data{ row.GetScrollbarData() };
+        if (data.has_value())
+        {
+            marks.emplace_back(y, *data);
+        }
+    }
+    return marks;
+}
+
+// Get all the regions for all the shell integration marks in the buffer.
+// Marks will be returned in top-down order.
+//
+// This possibly iterates over every run in the buffer, so don't do this on a
+// hot path. Just do this once per user input, if at all possible.
+//
+// Use `limit` to control how many you get, _starting from the bottom_. (e.g.
+// limit=1 will just give you the "most recent mark").
+std::vector<MarkExtents> TextBuffer::GetMarkExtents(std::optional<size_t> limit) const
 {
-    return _marks;
+    if (limit.has_value() &&
+        *limit == 0)

```

Comment:
fwiw for optionals you can directly compare them:

```suggestion
    if (limit == 0)
```

this will be false if `limit` is nullopt

Reply:
you can make a similar change below[^1] since you are checking `has_value` on every iteration of the loop.

Or, you could cache whether it has a value to avoid the constant checks

[^1]: though may need to read the docs on how `>=` works for nullopt

Reply:
> &nbsp; | &nbsp; | &nbsp;
> -- | -- | --
> `template< class T, class U >constexpr bool operator>=( const T& value, const optional<U>& opt );` | (32) | (since C++17)



<!--EndFragment-->
</body>
</html>

> 32) Returns bool(opt) ? value >= *opt : true.

https://en.cppreference.com/w/cpp/utility/optional/operator_cmp

Reply:
Darn, I guess that's the one we _don't_ want. At least equality works normally!